feat(spring-boot-starter): improve external task client starter

tasso94 opened the pull request:
* Incorporate community extension into mono repo (CAM-13126)
* Remove Lombok (CAM-13133)
* Remove possibility to bootstrap multiple clients (CAM-13135)
* Streamline codebase (CAM-13131)
  * Add Spring Boot banner
  * Add Javadocs
  * Reorganise package structure
  * Improve logging
  * Annotation attribute values are nullable where possible
* Create feature parity with the Java External Task Client (CAM-13149)
* Emit spring event when a subscription is initialized (CAM-13170)
* Streamline annotations and annotation behavior (CAM-13169)
* Improve mapping of annotations to configuration objects (CAM-13171)
* Streamline registration of client and subscription beans (CAM-13254)
* Increase test coverage (CAM-13127)

related to CAM-10863

tasso94 commented on spring-boot-starter/starter-client/spring/src/main/java/org/camunda/bpm/client/spring/impl/client/ClientRegistrar.java:
```diff
@@ -0,0 +1,76 @@
+
+  protected void registerExternalTaskClient(AnnotationMetadata importingClassMetadata,
+                                            BeanDefinitionRegistry beanDefinitionRegistry) {
+    if (beanDefinitionRegistry.containsBeanDefinition(BEAN_NAME)) {
```
We should make this check independent of the bean name `externalTaskClient` since a bean with a different name exposed by the user should also skip the external task bean creation.

tmetzke replied:
I agree 👍

tmetzke replied:
Definitely 👍 😃

tasso94 replied:
👍

tasso94 commented on spring-boot-starter/starter-client/spring/src/main/java/org/camunda/bpm/client/spring/impl/subscription/SubscriptionPostProcessor.java:
```diff
@@ -0,0 +1,106 @@
+import org.springframework.beans.factory.support.BeanDefinitionRegistryPostProcessor;
+import org.springframework.core.type.AnnotatedTypeMetadata;
+
+public class SubscriptionPostProcessor implements BeanDefinitionRegistryPostProcessor {
```
```suggestion
public class SubscriptionPostProcessor implements ImportBeanDefinitionRegistrar {
```
I think we should use `ImportBeanDefinitionRegistrar` instead of `BeanDefinitionRegistryPostProcessor`. While it seems to work, `BeanDefinitionRegistryPostProcessor` is not meant to be imported via annotation but rather exposed as a dedicated bean picked up by spring.

tmetzke replied:
That sounds reasonable 👍

tmetzke replied:
That sounds reasonable 👍

tasso94 replied:
For the record, we decided the following:
* Changing the `BeanDefinitionRegistryPostProcessor` to `ImportBeanDefinitionRegistrar` does not work since the latter has only class-level beans in the bean definition registry and not method-level beans which is a requirement to register subscription beans in a sensible way
* Since`BeanDefinitionRegistryPostProcessor` is more powerful and more commonly used compared to `ImportBeanDefinitionRegistrar`, we decided to convert the `ClientRegistry` into a `BeanDefinitionRegistryPostProcessor`
* Both post processors are registered at a central point in `PostProcessorConfiguration` and imported via `@Import` annotation in the `@EnableExternalTaskClient` annotation
  * Whenever the `@EnableExternalTaskClient` is added to a class, the `ClientPostProcessor` as well as `SubscriptionPostProcessor` are activated

tasso94 replied:
Documented in CAM-13254

tmetzke commented on spring-boot-starter/starter-client/spring-boot/src/main/java/org/camunda/bpm/client/spring/boot/starter/impl/properties/BasicAuthProperties.java:
```diff
@@ -0,0 +1,49 @@
+
+public class BasicAuthProperties {
+
+  protected boolean enabled = false;
```
With regards to other starters that we feature, we usually offer "configured == enabled" and "not configured == disabled" for authentication. I do not see a use case for this flag to be honest since you can also comment the configured authentication in the properties, so I would vote to remove it.
If we remove it, we need to make sure we handle them correctly in `PropertiesAwareClientFactory` in case of the properties not being defined in the `application.yml`

tasso94 replied:
👍

tasso94 commented on spring-boot-starter/starter-client/spring/src/main/java/org/camunda/bpm/client/spring/impl/client/ClientFactory.java:
```diff
@@ -0,0 +1,165 @@
+  public ExternalTaskClient getObject() {
+    if (client == null) {
+      ExternalTaskClientBuilder clientBuilder = ExternalTaskClient.create()
+          .baseUrl(resolve(clientConfiguration.getBaseUrl()))
+          .workerId(resolve(clientConfiguration.getWorkerId()));
```
Let's also skip calling these two builder methods when not configured.

tmetzke replied:
I agree 👍

tmetzke commented on spring-boot-starter/starter-client/spring/src/main/java/org/camunda/bpm/client/spring/impl/subscription/SubscriptionPostProcessor.java:
```diff
@@ -0,0 +1,106 @@
+          subscriptionConfiguration.fromAnnotation(subscriptionAnnotation);
+          BeanDefinition subscriptionBeanDefinition = getBeanDefinition(handlerBeanName, subscriptionConfiguration);
+
+          String subscriptionBeanName = handlerBeanName + "Subscription";
```
Alternative proposal:
* Let's use the topic name as the name for the subscription bean name so users can easily `@Autowired` those in their applications
* Let's add a defined suffix to the handler bean name and re-register the handler bean with that name, i.e. remove the old one and register it under the new name
* I would assume people do auto-wire the subscription bean more often to interact with it instead of the handler bean, so let's make it easier for them 😃

tasso94 replied:
👍

tasso94 replied:
I agree with the topic name as the bean name for subscriptions. However, I would propose to keep the handler naming as is. Here is why:
* A user defines the handler bean explicitly via {{@Bean}} annotation, so the user can fully control the handler bean name on its own
* When having spring support in your IDE (like in IntelliJ), you can navigate to the bean definition by clicking on `@Qualifier("handler-name")` which does not work if we add a suffix programmatically

tasso94 replied:
As discussed. Let's keep everything as it is. I will create a follow-up ticket where I document our discussion.

tasso94 replied:
I documented our discussion points and the decision in CAM-13255.

tmetzke commented on spring-boot-starter/starter-client/spring/src/main/java/org/camunda/bpm/client/spring/impl/client/ClientFactory.java:
```diff
@@ -0,0 +1,165 @@
+  }
+
+  @Autowired(required = false)
+  public void setClientBackoffStrategies(BackoffStrategy backoffStrategy) {
```
```suggestion
  public void setClientBackoffStrategy(BackoffStrategy backoffStrategy) {
```

tasso94 replied:
👍

tmetzke commented on spring-boot-starter/starter-client/spring/src/main/java/org/camunda/bpm/client/spring/annotation/EnableExternalTaskClient.java:
```diff
@@ -0,0 +1,154 @@
+@Import({ClientRegistrar.class, SubscriptionPostProcessor.class})
+public @interface EnableExternalTaskClient {
+
+  String STRING_NULL_VALUE = "$null$";
```
Since we already talked about it lately:
I think that is fine for a placeholder, shouldn't be too collision-prone in my view. 👍

tasso94 replied:
👍

tmetzke commented on spring-boot-starter/starter-client/spring/src/main/java/org/camunda/bpm/client/spring/annotation/EnableExternalTaskClient.java:
```diff
@@ -0,0 +1,154 @@
+public @interface EnableExternalTaskClient {
+
+  String STRING_NULL_VALUE = "$null$";
+  long LONG_NULL_VALUE = -1;
+  int INT_NULL_VALUE = -1;
```
Should we rather use some of those `Long#MIN_VALUE` and `Integer#MIN_VALUE` as placeholders as they are maybe less likely to collide with what users put in here? I could rather imagine someone putting in  a `-1` in the configuration rather than those other values and we would then be able to do a better validation maybe, telling them that their value of `-1` is not appropriate?
It is rather theoretical though, I admit. Feel free to reject it as well. 👍

tmetzke reviewed (commented):
Looks pretty damn good overall 👍 
Some smaller suggestions and ideas from my side, nothing critical really.
Some improvement ideas from your side already as well, so, yeah, I guess we're almost there already! 😃

tasso94 commented on spring-boot-starter/starter-client/README.md:
```diff
@@ -0,0 +1,69 @@
+# Camunda External Task Client as Spring Boot Starter
+
+This project provides a Spring Boot starter that allows you to implement an external task worker for Camunda. It uses the Camunda REST API to fetch, lock and complete external services tasks. 
+It is based on the [Camunda External Task Client](https://github.com/camunda/camunda-bpm-platform/clients/java)
```
👍

tasso94 commented:
@tmetzke, additionally to what we have discussed, I improved the exception handling slightly. Please also see the spring-only test cases. Next, I'll tackle the spring-boot tests.

tmetzke reviewed (commented):
That looks absolutely fabulous so far! 🚀 👍 
I might have missed it, but do we have a test case for injecting a subscription bean based on the implicit naming after the handler bean already? Should we include one as a) an example and b) a verification that this works as expected?

tasso94 commented:
Thanks for the trailblazing fast review! 🚀 👍 

We have a [test](
https://github.com/camunda/camunda-bpm-platform/blob/0c7d2210bd3e798171ef1608d4d91ba8d877256e/spring-boot-starter/starter-client/spring/src/test/java/org/camunda/bpm/client/spring/subscription/SubscriptionTest.java#L37). However, I think a more simple test that could also serve as an example would probably be nice.

tmetzke commented:
> We have a [test](https://github.com/camunda/camunda-bpm-platform/blob/0c7d2210bd3e798171ef1608d4d91ba8d877256e/spring-boot-starter/starter-client/spring/src/test/java/org/camunda/bpm/client/spring/subscription/SubscriptionTest.java#L37). However, I think a more simple test that could also serve as an example would probably be nice.

There it is, I couldn't find it again 😄 I think it is totally fine also as an example 👍

tasso94 commented:
@tmetzke, could you please have a look again? I added the missing spring boot tests and improved some smaller things...
* I added dependency management for powermock among all spring boot modules
* Found a problem with the new exception handling when a subscription is closed by spring
* Two tiny improvements

tasso94 commented on spring-boot-starter/starter-client/spring-boot/src/main/java/org/camunda/bpm/client/spring/boot/starter/impl/properties/ClientProperties.java:
```diff
@@ -0,0 +1,56 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.camunda.bpm.client.spring.boot.starter.impl.properties;
```
Let's move the property classes to `org.camunda.bpm.client.spring.boot.starter` and get rid of `impl.properties`. I think it is common practice to have the property classes in the top-level package.
Also, it looks somehow forbidden when a user auto-wires the properties from an `impl` package.

tmetzke replied:
I agree with both arguments 👍

tmetzke reviewed (approved):
The new tests and smaller improvements look good to me 👍

tmetzke commented:
🚀
